src/treebuilders/multiply.h

```diff
@@ -28,57 +28,65 @@
 #include "trees/FunctionTreeVector.h"
 
 namespace mrcpp {
-template <int D> class RepresentableFunction;
-template <int D> class FunctionTree;
+template <int D, typename T> class RepresentableFunction;
+template <int D, typename T> class FunctionTree;
 
-template <int D> void dot(double prec,
-                          FunctionTree<D> &out,
-                          FunctionTreeVector<D> &inp_a,
-                          FunctionTreeVector<D> &inp_b,
+template <int D, typename T> void dot(double prec,
+                          FunctionTree<D, T> &out,
+                          FunctionTreeVector<D, T> &inp_a,
+                          FunctionTreeVector<D, T> &inp_b,
                           int maxIter = -1,
                           bool absPrec = false);
 
-template <int D> double dot(FunctionTree<D> &bra,
-                            FunctionTree<D> &ket);
+template <int D, typename T> T dot(FunctionTree<D, T> &bra,
+                            FunctionTree<D, T> &ket);
 
-template <int D> double node_norm_dot(FunctionTree<D> &bra,
-                                      FunctionTree<D> &ket,
+template <int D> ComplexDouble dot(FunctionTree<D, ComplexDouble> &bra,
+                            FunctionTree<D, double> &ket);
+template <int D> ComplexDouble dot(FunctionTree<D, double> &bra,
+                            FunctionTree<D, ComplexDouble> &ket);
```

Comment:
I think it's enough to have one function, but with three template parameters (similarly could be done for all arithmetic functions, actually)
```suggestion
template <int D, typename T, typename U = T, typename V = decltype(std::declval<T>() * std::declval<U>())>
V dot(FunctionTree<D, T> &bra, FunctionTree<D, U> &ket);
```
you can see why I dislike implementing templates in `.cpp` files :sweat_smile: (cough cough @ilfreddy)

Note that in the C++ standard mixing `float` and `std::complex<double>` (or `double` and `std::complex<float>`) is _not_ possible, so the above should be combined with a compile-time exclusion of such a possibility (through SFINAE or something) Another thing I have in my archives.

Reply:
This is nice!